Initial Version

rgaudin opened the pull request:
- Basic functionnality present
- Works both standalone and all-in-one Docker image

@benoit74 that's basically a transfer of what was on kci and running on kathrin.
Please raise your questions and I'll document whatever needs to be.

rgaudin committed:
Initial Version

- Basic functionnality present
- Works both standalone and all-in-one Docker image

benoit74 commented on .github/workflows/DockerCD.yaml:
```diff
@@ -0,0 +1,25 @@
+
+jobs:
+
+  bittorrent-tracker:
```
`bittorent-tracker` => `bittorent-seeder`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+[![Docker](https://ghcr-badge.egpl.dev/kiwix/bittorrent-seeder/latest_tag?label=docker)](https://ghcr.io/kiwix/bittorrent-seeder/)
+
+
+It  consists of a script that runs periodically and which consists mostly of:
```
extra space to remove

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+- Downloading the Kiwix OPDS Catalog
+- Matching its entries with your defined filters
+- Communicates with your qBittorrent instance (via HTTP)
```
Missing a line or two explaining that it will automatically download ZIMs matching the filter and delete older ZIMs

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+- Downloading the Kiwix OPDS Catalog
+- Matching its entries with your defined filters
+- Communicates with your qBittorrent instance (via HTTP)
+  - Removes out-of-date (dropped from Catalog) ZIMs from qBittorrent
```
It also removes ZIMs which do not match the filters anymore, so I would rephrase with something like `Removes unselected (out-of-date, dropped from Catalog) ZIMs from qBittorent (which in turns removes them from filesystem)`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+**Key features:**
+
+- Very easy to use
+- Very flexible filters so you can precisely select what to seed
```
=> what to download and seed

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+- Very easy to use
+- Very flexible filters so you can precisely select what to seed
+- Compatible with your existing qBittorrent (doesnt mess with your stuff)
```
`doesnt` => `doesn't`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+> [!CAUTION]
+> The parameters/config passed to `kiwix-seeder` is an indication of the new requested state.
+> Say you were using it and are seeding 20 torrents, if you relaunch it with filters that match only a single ZIM, **it won't add this ZIM to your list**, it will remove all thoe others (see `--keep` below) and add that one.
```
I don't get it, you say that `it won't add this ZIM to your list` and it will `add that one`, this seems contradictory.

benoit74 replied:
`thoe` => `the`

benoit74 replied:
`--keep` is not documented below as indicated, but only in the usage

rgaudin replied:
re-read the sentence ; it's clear to me. I want reader to understand that filters does not mean that a new run adds on top of existing collection but replaces it completely.
I've added some details but I'm not confident it's quite clear.

benoit74 commented on src/kiwixseeder/entrypoint.py:
```diff
@@ -0,0 +1,304 @@
+The Catalog only exposes the latest version of a *Title*. With many *Titles* being \
+redone monthly.\n
+The webseeds (HTTP mirrors) only keep at most 2 versions of a *Title*.
+This option expects a
```
Misses end of the phrase.

benoit74 replied:
You should probably explain here (or somewhere else) that ZIMs which do not match the `--keep` criteria are both removed from qBittorrent and from the local filesystem.

benoit74 commented on src/kiwixseeder/entrypoint.py:
```diff
@@ -0,0 +1,304 @@
+## Keeping ZIMs
+
+`--keep` allows you to keep seeding ZIMs after those drop out of the Catalog.
+The Catalog only exposes the latest version of a *Title*. With many *Titles* being \
```
`With many *Titles* being redone monthly` => `Mostly all *Titles* are redone at least quarterly, many are redone monthly.`

benoit74 commented on README.md:
````diff
@@ -1,2 +1,190 @@
+This version depends on a reachable qBittorrent instance. You call it to update your qBittorrent's list of ZIMs to seed.
+
+```sh
+❯ kiwix-seeder --lang bam --max-storage 1GB
````
You should probably add here as well how we are supposed to configure qBittorrent URL (`export QBT_URL="http://admin:mypass@nas.local:8080"` if I'm not mistaken), since it will be needed in most cases.

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+### Docker version
+
+The docker version includes qBittorrent so it's meant to run forever.
```
`docker` => `Docker`

benoit74 replied:
`forever` => `as a daemon`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+The docker version includes qBittorrent so it's meant to run forever.
+
+The assistant script starts it from your config file and periodically runs the `kiwix-seeder` script to manage the list of torrents.
```
`The assistant script starts it from your config file and periodically runs` => `An assistant script is provided in the Docker image as entrypoint. It will handle starting qBittorrent instance from your config file and periodically running

benoit74 commented on README.md:
````diff
@@ -1,2 +1,190 @@
+❯ docker stop seeder
+```
+
+You can also run it without the helper script using image `ghcr.io/kiwix/bittorrent-seeder:latest`. You'll have to sort out volume mounting, port forwarding and configuration. Use the helper script as documentation!.
````
`as documentation!` => `as documented`

rgaudin replied:
No, you misread that. If you dont want the script, use it as documentation to set flags manually

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+| Standalone Binary | If you already have a running qBittorrent instance. | Lightweight and flexible |
+| Docker Image      | All in one docker image that runs both the script and qBittorrent. | Simplest  |
+
+The Docker version obviously depends on Docker being installed and running but
```
Missing end of phrase + worth mentioning `Docker or any alternative container distribution like Rancher`

rgaudin replied:
Stopped sentence after running. I think this was started before I broke it into subsections. I dont want to mention Docker alternatives. It's just a README and that's implied to those knowledgeable IMO

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+### Docker version
+
+This version is intended for those who want a setup-and-forget solution. It comes with qbittorrent. There's a good number of options but it's tailored for a kiwix-seeder only usage so it's not as flexible (although you can change any seetings via the WebUI)
```
I don't get why it is not as flexible as the other option, maybe explain a bit more. Is it because we do not have access to the qBittorrent install?

`seetings` => `settings`

rgaudin replied:
It's because there's a million option in qbt. As indicated, one can still change everything using the WebUI.

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+❯ mkdir -p /data/seeder
+❯ cat <<EOF > /etc/seeder.config
+
+# where to store all data
```
add `(ZIMs, ...)`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+# where to store all data
+DATA_PATH=/data/seeder
+
+# webui password used by script to communicate with qBt and user on remove webui
```
`remove` => `remote`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+DATA_PATH=/data/seeder
+
+# webui password used by script to communicate with qBt and user on remove webui
+QBT_PASSWORD="Choose this one"
```
Is it possible to not store password inside a config file? (I guess, probably not, but better to ask ^^)

rgaudin replied:
Not for this, no. We could allow a secret file instead art some point. Password is stored encrypted in qbt password and we cant decrypt that so changing it in webui prefs would break communication

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+# webui password used by script to communicate with qBt and user on remove webui
+QBT_PASSWORD="Choose this one"
+
+# BT port to use/announce. **must** be forwarded on router to local IP (uPNP cannot work accross docker routing)
```
`on router to local IP` => `from your public IP (for home users, probably to be done on your ISP router)`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+QBT_TORRENTING_PORT=6901
+
+# port on host to map webui (for remote access, optional)
+WEBUI_PORT=8080
```
Would it be possible to easily expose WEBUI_PORT only on `localhost`? Would be greatly superior in terms of security, especially for those who will run this from a VPS.

rgaudin replied:
See #2 which is exactly that: not binding to a port on the host. Would not allow user to access webui though.

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+# port on host to map webui (for remote access, optional)
+WEBUI_PORT=8080
+
+# max storage to use
```
`max storage to use` => `max size of DATA_PATH folder before ...` (what happens when this max is reached?)

rgaudin replied:
Currently (initial version shortcut), it doesn't refresh

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+# max storage to use
+MAX_STORAGE="10TiB"
+
+# interval between catalog refreshes
```
`catalog refreshes` => `kiwix-seeder invocations (catalog refresh, ZIM addition/removal)`

benoit74 commented on README.md:
````diff
@@ -1,2 +1,190 @@
+❯ chmod +x /usr/local/bin/seeder-start-restart
+```
+
+That's it. If you created the config file in a different place, edit the helper script to load it properly.
````
Shouldn't we at least start the helper script? Or even put it in a systemd-unit for automatic restart when the machine starts.

rgaudin replied:
Added start advice. Automatic restart is already taken care of by `--restart unless-stopped` on `docker run`.

benoit74 commented on README.md:
````diff
@@ -1,2 +1,190 @@
+Simply download and invoke it
+
+```sh
+❯ curl -o /usr/local/bin/kiwix-seeder https://TBD && chmod +x /usr/local/bin/kiwix-seeder
````
What is TBD, what should it be replaced with? Is it necessary if QBT_URL env var is set as suggested below?

rgaudin replied:
TBD means To Be Determined. That's because we lack #1 so I dont know the URLs for the binary.

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+#### qBittorrent requirements
+
+- qBittorrent must be running when using `kiwix-seeder`. Once it has run, qBittorrent does not need to keep running.
```
`Once it has run, qBittorrent does not need to keep running.` => `Once invocation of kiwix-server has completed, qBittorrent can be stopped/started at your convenience.`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+- qBittorrent must be running when using `kiwix-seeder`. Once it has run, qBittorrent does not need to keep running.
+- WebUI must be enabled and configured (see below)
+- The machine running `kiwix-seeder` must be able to communicate with that URL. Check that you can make an HTTP request from that machine using curl to ensure WebUI is working, reachable and the credentials are correct:
```
- `with that URL` => `with qBittorrent web UI URL`
- `from that machine` => `from kiwix-seeder machine to qBittorent URL`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+Then when running, `kiwix-seeder` will remove all the torrents and their associated files.
+
+You can also do it outside of the tool, using qBittorrent UI or WebUI. Simply right-click on the `kiwix-seeder` category and select IRemove torrents. You'll be promptyed to confirm and whether you want to delete the associated files.
```
- `IRemove`? Is it a software from Apple? 🤣 
- `You'll be promptyed to confirm and whether` => `You'll be prompted to confirm whether`

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+
+## Configuration
+
+...
```
We should at least redirect users to `kiwix-seeder` usage (at least in a first version)

benoit74 commented on entrypoint.sh:
```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
+
+function configure_qbt {
+	QBITTORRENT_CONFIG_FILE=/root/.config/qBittorrent/qBittorrent.conf
```
The names `QBITTORRENT_CONFIG_FILE` and `QBT_CONFIG_FILE` are pretty confusing. Maybe rename `QBT_CONFIG_FILE` to `QBT_SETTINGS_FILE`, and then `QBITTORRENT_CONFIG_FILE` to `QBT_CONFIG_FILE` for shortness

rgaudin replied:
Sure is confusing because we have qBittorrent with binary `qbittorrent-nox` for which we expose stuff so we want short names and qBittorrent-CLI with binary `qbt` which is an extra with no specific option.

I've changed to `QBT_CONFIG_FILE` so it's like the others and `QBT_CLI_CONFIG_FILE`

benoit74 commented on pyproject.toml:
```diff
@@ -0,0 +1,247 @@
+authors = [
+  { name = "Kiwix", email = "dev@kiwix.org" },
+]
+keywords = ["offspot"]
```
?

rgaudin replied:
suggestions welcome :)

rgaudin replied:
```toml
keywords = ["kiwix", "bittorrent", "torrent", "seed"]
```

benoit74 commented on pyproject.toml:
```diff
@@ -0,0 +1,247 @@
+
+[project.scripts]
+kiwix-seeder = "kiwixseeder.entrypoint:entrypoint"
+kiwix-seeder-loop = "kiwixseeder.forever:entrypoint"
```
What is it? It is not mentioned anywhere in the README

rgaudin replied:
Not supposed to be in the README. It's just the script that is in the Dockerfile's `CMD`. Its just a forever-running script that calls `kiwix-seeder` periodically

benoit74 commented on pyproject.toml:
```diff
@@ -0,0 +1,247 @@
+
+[tool.black]
+line-length = 88
+target-version = ['py311']
```
py312

benoit74 commented on seeder-start-restart.sh:
```diff
@@ -0,0 +1,96 @@
+# FILTERS
+# - use glob-patterns
+# - split using pipe (|)
+# - combined with same filter
```
This is a bit confusing. I imagine we combine with `OR` on same kind, and combine various kinds with `AND`. Probably an example would help

benoit74 commented on seeder-start-restart.sh:
```diff
@@ -0,0 +1,96 @@
+# - use glob-patterns
+# - split using pipe (|)
+# - combined with same filter
+FILENAMES=""										# matching filename (filepath actually)
```
example filepath would help

benoit74 commented on seeder-start-restart.sh:
```diff
@@ -0,0 +1,96 @@
+# If using a remote qBittorrent instance, see NO_DEAMON below
+QBT_TORRENTING_PORT=6901                            # port to use for bittorrent. **MUST** be manually opened and forwarded to this host's IP as uPNP would not work accross docker
+QBT_PASSWORD=""                                     # qBittorrent WebUI password. If empty, one will be gen and printed
+WEBUI_PORT=8000                                     # port on this host to map to the qBittorrent WebUI (so you can monitor it)
```
Default should probably be `127.0.0.1:8000` so that we expose WEBUI only on localhost, available only for kiwix-seeder operation by default.

rgaudin replied:
As explained above, #2 will allow this but I dont think it should be the default. WebUI is extremely useful to oversee what's happening and binding to localhost would prevent it from being accessed by the user.

benoit74 commented on src/kiwixseeder/utils.py:
```diff
@@ -0,0 +1,66 @@
+        """whether value is lower-or-equal than our maximum"""
+        if self.maximum == nd:
+            return True
+        return value <= self.maximum
```
Maybe `return self.maximum == nd or value <= self.maximum` is more concise and sufficiently clear?

benoit74 commented on src/kiwixseeder/entrypoint.py:
```diff
@@ -0,0 +1,304 @@
+    DEFAULT_QBT_CONN,
+    Context,
+    QbtConnection,
+    SizeRange,
```
`SizeRange` comes from `kiwixseeder.utils` in fact (even if it works)

rgaudin replied:
it's incidental (used to be in context)

benoit74 commented on README.md:
```diff
@@ -1,2 +1,190 @@
+- Matching its entries with your defined filters
+- Communicates with your qBittorrent instance (via HTTP)
+  - Removes out-of-date (dropped from Catalog) ZIMs from qBittorrent
+  - Adds new matching ZIM to qBittorrent
```
`(which in turns will download them to the filesystem)`

rgaudin replied:
You already reported that 😉

benoit74 commented on src/kiwixseeder/context.py:
```diff
@@ -0,0 +1,191 @@
+    os.getenv("KEEP_DURATION") or "12w"
+)
+DEFAULT_SLEEP_INTERVAL: float = humanfriendly.parse_timespan(
+    os.getenv("SLEEP_INTERVAL") or "5m"
```
Idem, bigger default ?

benoit74 commented on seeder-start-restart.sh:
```diff
@@ -0,0 +1,96 @@
+
+DATA_PATH=$(pwd)/kiwix-seeder                       # path to store ZIM files (and incomplete .!qB ones) in (there's no hierarchy)
+MAX_STORAGE="10GiB"									# maximum disk-space to use
+SLEEP_INTERVAL="5m"                                 # how long to pause in-between catalog checks when using --loop
```
Shouldn't we have a bigger sleep interval by default ? I feel like 1h is already a lot, maybe even 1 day should be advised as a default value to avoid flooding our servers for a very limited added value (geeks always want the most up-to-date information, so let's try to avoid pushing in their direction ^^).

benoit74 reviewed (changes requested):
Very good work obviously. 

Of course I have found lot's of things to comment, but nothing very big.

I think the program overall misses few safeguards:
- avoid silly (small) settings of sleep period (e.g. never smaller than 5 mins) to avoid flooding our servers
- avoid removing too many ZIMs at once => this is a bit blurry, but the scenario I have in mind is that should we publish an empty catalog for any reason (or should the program believe the catalog is empty for any reason), it would be a pity / concern to cause all ZIMs of all peers to be deleted within few hours ; maybe even deleting too many ZIMs at once is an indicator of an issue ; that been said, since we keep the ZIMs for `--keep` days, the impact is obviously not immediate ... should we simply have an alerting system? a summary file one can watch with details about files which are going to be removed in 1 week, 2 weeks, 4 weeks, 8 weeks, 12 weeks?

rgaudin committed:
Changed qbt and qbittorrent var names for config files in entrypoint

rgaudin committed:
Typos

rgaudin committed:
Slight improvements to filters:

- added Title and Description filters.
- Removed nodet flavour choice (now gone)
- All matches now case insensitive
- Default sleep interval to 1d for docker script

rgaudin committed:
Fixed #4: fixed usage in entrypoint

rgaudin committed:
Fixed #3: improved README

rgaudin committed:
Allow re-querying catalog in dry-run mode

rgaudin commented:
Thank you @benoit74 ; I've made all the necessary edits ; and rebased. I opened #8 regarding mass removal as I have no straight forward solution
